### PR TITLE
WIP: Prepare the ground for using a new calling convention for functions from the runtime library

### PR DESCRIPTION
**Pull request by swiftix**
This PR adds support for using a new calling convention for functions from a runtime library.
## What is the reason for using a new calling convention?

Some of the runtime functions, in particular reference-counting related functions, are very, very hot and have a huge number of call sites in a typical compiled Swift code. The idea is to use for these runtime functions a new calling convention which has more callee-saved registers than a standard calling convention. Using of this calling convention for runtime functions reduces the amount of spilling required on the caller side, allows for better register allocation and leads to the overall code size reduction.
## What is in this PR?

A big part of the this patch-set just prepares the machinery for enabling the use of different calling conventions with runtime functions. 

This includes among other things:
- implementing many fixes in IRGen and LLVMPasses, because they were not prepared for using different calling conventions. They silently ignored the actual calling convention defined in the RuntimeFunctions.def and generated code assuming that the standard C calling convention is used everywhere.
- supporting the generation of wrappers that should be used by the a client code to invoke functions from the runtime library.
  -   These wrappers are required as a workaround for issues related to usage of dynamic linker stubs.
    -  Typically, if a function from a dynamic library is invoked from the a client code, the emitted code would make use of dynamic linker stubs. These stubs would be lazily resolved at runtime, when the function is invoked for the first time. Unfortunately, when the new calling convention is used, the dynamic linker while performing the lazy dynamic binding may clobber some of the registers claimed as callee-saved registers by this calling convention. And this may result in an undefined behaviour or even application crash. Therefore, any lazy dynamic linking of runtime functions using the new calling convention should be avoided. This is achieved by (1) introducing in the runtime library a global variable, which a is function pointer referring to the implementation of a given runtime entry and (2) generating a wrapper which invokes the implementation of the runtime entry via an indirect call through this global variable.
  - Usage of wrappers instead of dynamic linker stubs may also improve performance in some cases. For example some non-insignificant improvements were observed by benchmarking some retain/release heavy workloads. This is due to the fact that the wrapper-based approach removes one level of indirection compared to the currently used approach.

Finally, this PR defines which runtime functions should use the new calling convention:
- annotating functions in the runtime library with a proper calling convention.
  - Many variants of retain/release functions are annotated to use the new calling convention.
    But those variants of retain/release functions that may result in calls of objc_retain or objc_release
    are not migrated to the new calling convention, because it results in significant performance degradations
    when objects of Obj-C derived classes are used.
  - Some popular non-reference counting functions like swift_getGenericMetadata or swift_dynamicCast are annotated as well.

Most of the changes in this PR are not tied to the specific calling convention and make it rather easy to add use new calling conventions in the future.
## Current status

Everything is prepared for the use of this new calling convention, but the calling convention itself is not enabled yet. This has to wait until the required changes for supporting this calling convention land in clang and llvm in the following weeks. Therefore, as of now, the standard C calling convention is actually used for all the runtime functions. But all the machinery including the use of wrappers is fully functional and can be used.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
I tried to split changes in relatively small commits, which should be hopefully easy to review.

@jckarter Joe, could you review this PR? I don't know if you want to review the whole thing, but at least the runtime library related changes? The IRGen changes are actually not that big. It is mostly the generation of wrappers and fixing the propagation of the calling convention defined in RuntimeFunctions.def into call instructions.

**Comment by jckarter** [comment]
Looks good, aside from my naming concerns.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@jckarter Joe, please have a look. I updated the naming based on your feedback. 

I also added a simple test. It will get more interesting once we really enable the new calling convention.

**Comment by jckarter** [comment]
Looks good.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
Trying to figure out why Linux build times out.

**Comment by swiftix** [comment]
@jckarter Hey Joe, do you mind taking another look at this PR? 

It is a follow-up of what you have reviewed already. 

I finally managed to make it run under Linux. There was a very subtle issue related to a dynamic symbol lookup in interpreter mode on ELF-based systems! ;-)

Other than that, I figured out that it is better if all those generated wrappers have a prefix to distinguish them from the real runtime functions (before this change wrappers had the same names as runtime functions, but a hidden visibility. But there were some issues with this in the interpreter mode). I picked the `rt_` prefix, i.e. it now looks like `rt_swift_retain`. If we want to always have `swift` at the beginning, I can change it to become a prefix, so that we get 'swift_retain_rt'.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Looks good @swiftix.

**Comment by swiftix** [comment]
@rjmccall Do you mind having a look at this PR? It introduces some changes to IRGen:
 - RuntimeFunctions.def is moved into a place where it could be shared by other parts of the compiler.
 - Better propagation of calling conventions down to the llvm Call instructions.
 - Generation of wrappers for invocations of runtime functions using the new calling convention. This is what we discussed last week.

And of course it has all the required changes in the runtime library to make use of the new calling convention.

**Comment by swiftix** [comment]
@swift-ci Please test OS X platform

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please test

File: stdlib/public/runtime/HeapObject.cpp [hunk 1]
```diff
@@ -572,29 +653,31 @@ void swift::swift_deallocObject(HeapObject *object, size_t allocatedSize,
...
 
 HeapObject *swift::swift_weakLoadStrong(WeakReference *ref) {
   auto object = ref->Value;
   if (object == nullptr) return nullptr;
   if (object->refCount.isDeallocating()) {
-    swift_unownedRelease(object);
+    SWIFT_RT_ENTRY_CALL(swift_unownedRelease)(object);
     ref->Value = nullptr;
     return nullptr;
   }
```
Comment (glessard) [inline stdlib/public/runtime/HeapObject.cpp]: Should `swift_tryRetain` (on line 684) also be a SWIFT_RT_ENTRY_CALL ?

Comment (jckarter) [inline stdlib/public/runtime/HeapObject.cpp]: I think Instruments only needs to intercept the primitive refcounting operations.

**Pull request merged**